File: Maxwell_Early_Closed_Loop_Tests/closed_loop_2/edited_code/close_loop.py
import sys
import zmq
import struct
import array
from collections import namedtuple


SpikeEvent = namedtuple('SpikeEvent', 'frame channel amplitude')


def send_command(api_socket, string):
    '''
    Send a MaxWell API command to a ZMQ socket, check that the reply is
    "OK", and if not, print the error that occurred.
    '''
    api_socket.send_string(string)
    reply = api_socket.recv_string()
    if reply != 'OK':
        print(f'An error occurred {reply}.')
        api_socket.send_string('get_errors')
        print(api_socket.recv_string())

 
if __name__ == '__main__':
    ctx = zmq.Context.instance()

    # Provide an argument to do closed loop on that channel, otherwise
    # don't do any stim but print spikes as they happen.
    receive_channel = -1 if len(sys.argv) < 2 else int(sys.argv[1])

    # Generate the subscriber sockets with the same settings as the
    # original C++ program.
    subscriber = ctx.socket(zmq.SUB)
    subscriber.setsockopt(zmq.RCVHWM, 0)
    subscriber.setsockopt(zmq.RCVBUF, 10*20000*1030)
    subscriber.setsockopt_string(zmq.SUBSCRIBE, "")
    subscriber.setsockopt(zmq.RCVTIMEO, 100)
    subscriber.connect('tcp://localhost:7205')

    # Also generate a similar API socket.
    api_socket = ctx.socket(zmq.REQ)
    api_socket.setsockopt(zmq.RCVTIMEO, -1)
    api_socket.connect('tcp://localhost:7210')

    # This first loop ignores any partial packets to make sure the real
    # loop gets aligned to an actual frame. First it spins for as long
    # as recv() fails, then it waits for the RCVMORE flag to be False to
    # check that the last partial frame is over.
    more = True
    while more:
        try:
            msg = subscriber.recv()
        except zmq.ZMQError as e:
            continue
        more = subscriber.getsockopt(zmq.RCVMORE)

    while True:
        # Sometimes the publisher will be interrupted, so don't let that
        # crash the entire program, just skip the frame.
        frame_number = frame_data = events_data = None
        try:
            # The first component of each message is the frame number as
            # a long long, so unpack that.
            frame_number = struct.unpack('Q', subscriber.recv())[0]

            # We're still ignoring the frame data, but we have to get it
            # out from the data stream in order to skip it.
            if subscriber.getsockopt(zmq.RCVMORE):
                frame_data = subscriber.recv()

            # This is the one that stores all the spikes.
            if subscriber.getsockopt(zmq.RCVMORE):
                events_data = subscriber.recv()

        except Exception as e:
            print(e)
            continue

        # `frame_data` is a 1027-element array containing the recorded voltage
        # at each electrode, so unpack that into a usable format.
        frame = array.array('f', frame_data)

        events = []
        if events_data is not None:
            # The spike structure is 8 bytes of padding, a long frame
            # number, an integer channel (the amplifier, not the
            # electrode), and a float amplitude.
            spike_struct = '8xLif'
            spike_size = struct.calcsize(spike_struct)
            if len(events_data) % spike_size != 0:
                print(f'Events has {len(events_data)} bytes,',
                      f'not divisible by {spike_size}', file=sys.stderr)

            # Iterate over consecutive slices of the raw events
            # data and unpack each one into a new struct.
            for i in range(0, len(events_data), spike_size):
                ev = SpikeEvent(*struct.unpack(spike_struct,
                    events_data[i:i+spike_size]))
                events.append(ev)

                # If the event is on the channel of interest, tell the
                # server to send the stimulation pattern that gets
                # defined by the provided Python script.
                if ev.channel == receive_channel:
                    print(f'Event detected on {receive_channel}. Sending stim command...')
                    send_command(api_socket, 'sequence_send close_loop')


        # Just dump them to the console to show it works.
        if len(events) > 0:
            print(events)

    # If you split this code into multiple threads, e.g. to consolidate
    # this script with the one that handles the recordings, it will be
    # necessary to delete each subscriber to the ZMQ context separately
    # using the usual Python operator `del`, and then terminate the
    # context with `ctx.term()`. In a single-threaded script like this,
    # though, it's fine to just let the Python interpreter run all these
    # destructors when the program terminates.

File: Maxwell_Early_Closed_Loop_Tests/closed_loop_2/edited_code/close_loop.cpp
#include <vector>
#include <iostream>
#include <zmq.hpp>


struct SpikeEvent
{
    unsigned long filler = 0; // Added by me to try to make the representation make sense...
    unsigned long frameNo;    //!< Frame number (time point) at which spike is detected.
    int channel;              //!< Channel number at which spike is detected.
    float amp;                //!< Amplitude of a detected spike.

    SpikeEvent(unsigned long f, int c, float a)
        :frameNo(f),
        channel(c),
        amp(a)
    { }

    static const unsigned int bSize = 2*sizeof(unsigned long) +
                                      sizeof(int) +
                                      sizeof(float);

    SpikeEvent(const SpikeEvent& se)
        :frameNo(se.frameNo),
        channel(se.channel),
        amp(se.amp)
    {}

    SpikeEvent(unsigned char *buffer )
        : frameNo(*reinterpret_cast<unsigned long int*>(&buffer[8])),
        channel(*reinterpret_cast<int*>(&buffer[8+sizeof(frameNo)])),
        amp(*reinterpret_cast<float*>(&buffer[8+sizeof(frameNo)+sizeof(channel)]))
    { }
};

class EventDeserialize
{
    std::vector<SpikeEvent> spikes;
    public:
    EventDeserialize( unsigned char* buffer, unsigned int buffer_size )
    {
        for (unsigned int i=0 ; i < buffer_size ; i+=SpikeEvent::bSize )
        {
            spikes.push_back( SpikeEvent( &buffer[i] ) );
        }
    }
    const std::vector<SpikeEvent> &get_spikes() {return spikes;}
    unsigned int size(){return spikes.size();}
};


std::string send_command(zmq::socket_t *socket, std::string cmd)
{
    zmq::message_t msg(cmd.size());
    memcpy(msg.data(), cmd.data(), cmd.size());
    socket->send(msg);

    // receive back answer
    zmq::message_t msg_rdy;
    socket->recv(&msg_rdy);
    return std::string(static_cast<char*>(msg_rdy.data()), msg_rdy.size());
}

void dataRead(int detection_channel) {

    zmq::context_t *context = new zmq::context_t(1);

    // Socket to receive data from the server
    zmq::socket_t subscriber(*context, ZMQ_SUB);

    unsigned int val = 0;
    subscriber.setsockopt(ZMQ_RCVHWM, &val, sizeof(val));

    val = 10 * 20000 * 1030;
    subscriber.setsockopt(ZMQ_RCVBUF, &val, sizeof(val));

    // Subscribe to everything
    const char *filter = "";
    subscriber.setsockopt(ZMQ_SUBSCRIBE, filter, strlen(filter));

    // set timeout
    int timeout = 100;
    subscriber.setsockopt(ZMQ_RCVTIMEO, &timeout, sizeof(timeout));

    subscriber.connect("tcp://localhost:7205");


    // Socket to communicate with the API
    std::string connection = "tcp://localhost:7210";
    zmq::socket_t *api_socket = new zmq::socket_t(*context, ZMQ_REQ);

    timeout = -1;
    api_socket->setsockopt(ZMQ_RCVTIMEO,  &timeout, sizeof(timeout) );
    api_socket->connect ( connection.c_str() );

    
    // sync with the RCVMore mode stream
    while (1) {
        zmq::message_t message;
        try {
            subscriber.recv(&message);
        } catch (...) {
            continue;
        }
        int64_t more = 0; //  multipart detection
        size_t more_size = sizeof(more);
        subscriber.getsockopt(ZMQ_RCVMORE, &more, &more_size);
        if (!more)
            break;
    }

    unsigned long long blanking = 0;

    while (true) {
        zmq::message_t frameNo;
        zmq::message_t frame;
        zmq::message_t events;

        try {
            /* In case the publisher dies, make sure to synchronize again */
            subscriber.recv(&frameNo);

            int64_t more = 0;
            size_t more_size = sizeof(more);

            subscriber.getsockopt(ZMQ_RCVMORE, &more, &more_size);
            if (more)
                subscriber.recv(&frame);

            subscriber.getsockopt(ZMQ_RCVMORE, &more, &more_size);
            if (more) {
                subscriber.recv(&events);
            }
        } catch (...) {
        }

        unsigned long long frame_no = 0;
        memcpy((void *)&frame_no, (void *)frameNo.data(), 6);

        // data is equivalent to: float data[frame.size() / sizeof(float)]
        float *data = (float*) frame.data();

        if (blanking)
            blanking--;

        EventDeserialize e(static_cast<unsigned char*>(events.data()), events.size());
        const std::vector<SpikeEvent> & spikes = e.get_spikes();
        for (int i=0;i<spikes.size();i++) {
            //std::cerr << i << " Spike on channel " << spikes[i].channel 
            //    << " with amplitude " << spikes[i].amp << "\n";

            int chan = spikes[i].channel;
            if (chan == detection_channel && blanking==0)
            {
                std::cout << "Saw a spike on " << detection_channel << std::endl;
                std::string reply = send_command(api_socket, "sequence_send close_loop");

                blanking = 200 * 20;// 100ms

                if (reply!="OK") {
                    std::cerr << "An error occured: " << reply << "\n";
                    std::cerr << send_command(api_socket, "get_errors");
                }
            }
        }
    }
}



int main(int argc, char *argv[]) {
    if (argc != 2) {
        std::cerr << "Call with: " << argv[0] << " [detection_channel]" << std::endl;
        exit(1);
    }

    int detection_channel = atoi(argv[1]);

    dataRead(detection_channel);

    return 0;
}
